### PR TITLE
Meleespeech Action, Holoparasite Meleespeech

### PR DESCRIPTION
**Pull request by HerCoyote23**
## About the PR

Makes the MeleeSpeech component have an action to configure your battlecry.
Adds MeleeSpeech to Holoparasites.

Be advised that this is my first foray into actions and I was very much figuring out how they work over this PR. It's likely a bit flawed in implementation.

**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: The Syndicate's Holoparasite have taken to yelling catchphrases as they fight.

### CODE REVIEW

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 1]
```diff
@@ -1,49 +1,85 @@
...
-	/// <summary>
-	/// Attempts to change the battlecry of an entity.
-	/// Returns true/false.
-	/// </summary>
-	/// <remarks>
-	/// If provided with a player's EntityUid to the player parameter, adds the change to the admin logs.
-	/// </remarks>
-	public bool TryChangeBattlecry(EntityUid uid, string? battlecry, MeleeSpeechComponent? meleeSpeechComp = null)
+        if (!_uiSystem.TryToggleUi(storeEnt, MeleeSpeechUiKey.Key, actor.PlayerSession))
+            return;
```
Comment (EmoGarbage404) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:71]:
```suggestion
        _uiSystem.TryToggleUi(storeEnt, MeleeSpeechUiKey.Key, actor.PlayerSession);
```

no need to wrap it in a condition

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 2]
```diff
@@ -1,49 +1,85 @@
...
+    /// Attempts to open the Battlecry UI.
+    /// </summary>
+    private void OnConfigureAction(EntityUid uid, MeleeSpeechComponent comp, MeleeSpeechConfigureActionEvent args)
+    {
+
+        TryOpenUi(args.Performer, uid, comp);
+    }
 
 
+    public void TryOpenUi(EntityUid user, EntityUid storeEnt, MeleeSpeechComponent? component = null)
```
Comment (EmoGarbage404) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:62]: rename storeEnt (i see that you stole this from storesystem)

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs [hunk 3]
```diff
@@ -7,27 +8,45 @@ namespace Content.Shared.Speech.Components;
...
+        Priority = -20,
+        Event = new MeleeSpeechConfigureActionEvent(),
+    };
 }
 
 /// <summary>
 /// Key representing which <see cref="BoundUserInterface"/> is currently open.
 /// Useful when there are multiple UI for an object. Here it's future-proofing only.
-/// </summary>
+/// </summary>/
```
Comment (EmoGarbage404) [inline Content.Shared/Speech/Components/MeleeSpeechComponent.cs:38]:
```suggestion
/// </summary>
```

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs [hunk 4]
```diff
@@ -7,27 +8,45 @@ namespace Content.Shared.Speech.Components;
...
 	public string? Battlecry;
 
     [ViewVariables(VVAccess.ReadWrite)]
     [DataField("MaxBattlecryLength")]
     public int MaxBattlecryLength = 12;
+
+    [DataField("configureAction")]
+    public InstantAction ConfigureAction = new()
+    {
+        UseDelay = TimeSpan.FromSeconds(4),
```
Comment (EmoGarbage404) [inline Content.Shared/Speech/Components/MeleeSpeechComponent.cs:26]: this is a really long use delay considering it opens a ui window

**Review by EmoGarbage404 — CHANGES_REQUESTED** [review]

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs [hunk 5]
```diff
@@ -51,3 +70,13 @@ public MeleeSpeechBattlecryChangedMessage(string battlecry)
...
 }
+
+public sealed class MeleeSpeechConfigureActionMessage : InstantActionEvent
+{
+    public EntityUid User { get; }
+
+    public MeleeSpeechConfigureActionMessage(EntityUid who)
+    {
+        User = who;
+    }
```
Comment (keronshb) [inline Content.Shared/Speech/Components/MeleeSpeechComponent.cs:81]:
```suggestion
    public MeleeSpeechConfigureActionMessage(EntityUid user)
    {
        User = user;
    }
```

Small change but it's better to rename this for consistency.

**Review by keronshb — CHANGES_REQUESTED** [review]

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 6]
```diff
@@ -1,49 +1,84 @@
...
+    [Dependency] private readonly IAdminLogManager _adminLogger = default!;
+    [Dependency] private readonly SharedActionsSystem _actionSystem = default!;
+    [Dependency] private readonly UserInterfaceSystem _uiSystem = default!;
+    public override void Initialize()
 	{
 		base.Initialize();
 		SubscribeLocalEvent<MeleeSpeechComponent, MeleeSpeechBattlecryChangedMessage>(OnBattlecryChanged);
-	}
-
+        SubscribeLocalEvent<MeleeSpeechComponent, MeleeSpeechConfigureActionEvent>(OnConfigureAction);
```
Comment (metalgearsloth) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:22]: Please don't mix tabs and spaces, use spaces.

Also leave newlines between fields and methods.

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 7]
```diff
@@ -1,49 +1,84 @@
...
+    }
+    
+    private void OnGetActions(EntityUid uid, MeleeSpeechComponent component, GetItemActionsEvent args)
+    {
+        if (component.ConfigureAction != null)
+            args.Actions.Add(component.ConfigureAction);
+    }
 
 	private void OnBattlecryChanged(EntityUid uid, MeleeSpeechComponent comp, MeleeSpeechBattlecryChangedMessage args)
 	{
```
Comment (metalgearsloth) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:40]: Also here.

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 8]
```diff
@@ -1,49 +1,84 @@
...
 
         TryChangeBattlecry(uid, battlecry, meleeSpeechUser);
 	}
+    /// <summary>
+    /// Attempts to open the Battlecry UI.
+    /// </summary>
+    private void OnConfigureAction(EntityUid uid, MeleeSpeechComponent comp, MeleeSpeechConfigureActionEvent args)
+    {
+
+        TryOpenUi(args.Performer, uid, comp);
```
Comment (metalgearsloth) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:58]: Remove blank line.

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 9]
```diff
@@ -1,49 +1,84 @@
...
         if (battlecry.Length > comp.MaxBattlecryLength)
             battlecry = battlecry[..comp.MaxBattlecryLength];
 
         TryChangeBattlecry(uid, battlecry, meleeSpeechUser);
 	}
+    /// <summary>
+    /// Attempts to open the Battlecry UI.
+    /// </summary>
+    private void OnConfigureAction(EntityUid uid, MeleeSpeechComponent comp, MeleeSpeechConfigureActionEvent args)
+    {
```
Comment (metalgearsloth) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:56]: Also here.

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 10]
```diff
@@ -58,14 +93,12 @@ public bool TryChangeBattlecry(EntityUid uid, string? battlecry, MeleeSpeechComp
...
 
-		if (meleeSpeechComp.Battlecry == battlecry)
+        if (meleeSpeechComp.Battlecry == battlecry)
 
 			return true;
 
-
-
 		meleeSpeechComp.Battlecry = battlecry;
 		Dirty(meleeSpeechComp);
```
Comment (metalgearsloth) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:103]: Also here.

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 11]
```diff
@@ -58,14 +93,12 @@ public bool TryChangeBattlecry(EntityUid uid, string? battlecry, MeleeSpeechComp
 		else
 		{
 			battlecry = null;
-		}
+        }
 
-		if (meleeSpeechComp.Battlecry == battlecry)
+        if (meleeSpeechComp.Battlecry == battlecry)
 
 			return true;
```
Comment (metalgearsloth) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:100]: Also here.

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 12]
```diff
@@ -58,14 +93,12 @@ public bool TryChangeBattlecry(EntityUid uid, string? battlecry, MeleeSpeechComp
 		else
 		{
 			battlecry = null;
```
Comment (metalgearsloth) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:95]: Also here.

File: Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs [hunk 13]
```diff
@@ -1,49 +1,84 @@
...
+
+    /// <summary>
+    /// Attempts to change the battlecry of an entity.
+    /// Returns true/false.
+    /// </summary>
+    /// <remarks>
+    /// If provided with a player's EntityUid to the player parameter, adds the change to the admin logs.
+    /// </remarks>
+    public bool TryChangeBattlecry(EntityUid uid, string? battlecry, MeleeSpeechComponent? meleeSpeechComp = null)
 	{
```
Comment (metalgearsloth) [inline Content.Server/Speech/EntitySystems/MeleeSpeechSystem.cs:82]: Also here.

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs [hunk 14]
```diff
@@ -7,15 +8,28 @@ namespace Content.Shared.Speech.Components;
 [AutoGenerateComponentState]
 public sealed partial class MeleeSpeechComponent : Component
 {
+    [ViewVariables]
```
Comment (metalgearsloth) [inline Content.Shared/Speech/Components/MeleeSpeechComponent.cs:11]: Automatic component states don't actually work if the component isn't networked.

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs [hunk 15]
```diff
@@ -7,15 +8,28 @@ namespace Content.Shared.Speech.Components;
...
+    public EntityUid? User;
 
-	[ViewVariables(VVAccess.ReadWrite)]
+    [ViewVariables(VVAccess.ReadWrite)]
 	[DataField("Battlecry")]
 	[AutoNetworkedField]
 	public string? Battlecry;
 
     [ViewVariables(VVAccess.ReadWrite)]
```
Comment (metalgearsloth) [inline Content.Shared/Speech/Components/MeleeSpeechComponent.cs:19]: Don't mix tabs and spaces also please document your code, there's nothing on the system or component that explains what it actually does.

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs [hunk 16]
```diff
@@ -7,15 +8,28 @@ namespace Content.Shared.Speech.Components;
 [AutoGenerateComponentState]
 public sealed partial class MeleeSpeechComponent : Component
 {
+    [ViewVariables]
+    public EntityUid? User;
```
Comment (metalgearsloth) [inline Content.Shared/Speech/Components/MeleeSpeechComponent.cs:12]: Is this even used?

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs [hunk 17]
```diff
@@ -28,6 +42,11 @@ public enum MeleeSpeechUiKey : byte
     Key,
 }
 
+//[Serializable, NetSerializable]
+public sealed class MeleeSpeechConfigureActionEvent : InstantActionEvent
+{
+}
+
```
Comment (metalgearsloth) [inline Content.Shared/Speech/Components/MeleeSpeechComponent.cs:49]: huh?

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs [hunk 18]
```diff
@@ -51,3 +70,13 @@ public MeleeSpeechBattlecryChangedMessage(string battlecry)
...
+public sealed class MeleeSpeechConfigureActionMessage : InstantActionEvent
+{
+    public EntityUid User { get; }
+
+    public MeleeSpeechConfigureActionMessage(EntityUid user)
+    {
+        User = user;
+    }
+}
```
Comment (metalgearsloth) [inline Content.Shared/Speech/Components/MeleeSpeechComponent.cs:82]: Looks unused?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Speech/Components/MeleeSpeechComponent.cs (hunk 17 above)
Comment (HerCoyote23) [reply Content.Shared/Speech/Components/MeleeSpeechComponent.cs]: Afaik it's required for the event. Looking at MimePowers, it's put into the System instead, but both the component and system for MimePowers are in Server, where MeleeSpeechComponent is in Shared and System is in Server. I need to be able to access it in Component for the action definition to work.

I just made a bit less obtrusive in Component. Would it be better to put most of Component in Server and just have a few things in a separate SharedComponent?

**Comment by HerCoyote23** [comment]
I done goofed by comitting RobustToolbox then messing up the fix.

**Comment by HerCoyote23** [comment]
That may have been a mistake...

**Comment by HerCoyote23** [comment]
Couldn't save it, but I could replace it. New PR is up, #19504

**Pull request closed without merging**